tools/wake-format/emitter.cpp

```diff
@@ -201,23 +201,60 @@ static size_t count_trailing_newlines(const token_traits_map_t& traits, const CS
   return it->second.after_bound.size();
 }
 
-static size_t count_allowed_newlines(const token_traits_map_t& traits, const CSTElement& node) {
-  FMT_ASSERT(node.isNode(), node,
-             "Expected node, Saw <" + std::string(symbolName(node.id())) + ">");
-  return count_leading_newlines(traits, node) + count_trailing_newlines(traits, node);
+// Determines if a given node would emit a leading comment if emitted.
+static bool has_leading_comment(const CSTElement& node, const token_traits_map_t& traits) {
+  return count_leading_newlines(traits, node) > 0;
 }
 
-// Counts all newlines allowd before and after a require header. Ignores the require body
-// - require a = b # comment -> 1
+// Determines if a given node would emit a trailing comment if emitted.
+static bool has_trailing_comment(const CSTElement& node, const token_traits_map_t& traits) {
+  return count_trailing_newlines(traits, node) > 0;
+}
+
+// Determines if the doc is "weakly flat". Weakly flat is a flat doc
+// with a single trailing comment allowed. No other newlines my be emitted.
+static bool is_weakly_flat(const wcl::doc& doc, const CSTElement& node,
+                           const token_traits_map_t& traits) {
+  return !doc->has_newline() ||
+         (doc->newline_count() == 1 && count_trailing_newlines(traits, node) == 1);
+}
+
+// Determizes of a doc is "flat" as a human would judge it. Human's don't
+// consider leading/trailing comments as invalidating flatness even though they
+// require extra newlines. Thus leading/trailing newlines are allowed.
```

Comment:
I would change this comment to reflect the subtitles of what humans consider "flat", for instance this comment never matched my intuition. Something like this:

Determines if a doc is "vertically" flat. A vertically flat doc is "flat" if the only newlines in it come from comments. This is the notion of "flat" you would want to consider when arranging docs in a vertical list where only the "body" of the doc (e.g. not the leading or trailing comments) needs to be flat. Internal comments would however violate this property.